Implement list of available images matching logic

pgorszkowski-igalia opened the pull request:
#### 873115a0db7463b32eb5ef31965be9f8b21bcbb7
<pre>
Implement list of available images matching logic
 <a href="https://bugs.webkit.org/show_bug.cgi?id=243790">https://bugs.webkit.org/show_bug.cgi?id=243790</a>

Reviewed by Sihui Liu and Chris Dumez.

Implement list of available images matching logic:
<a href="https://html.spec.whatwg.org/multipage/images.html#updating-the-image-data">https://html.spec.whatwg.org/multipage/images.html#updating-the-image-data</a>:list-of-available-images

This is restricted to lazy loading for now.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/image-loading-lazy-available-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/image-loading-lazy-use-list-of-available-images-expected.txt:
* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::canReuseFromListOfAvailableImages):
(WebCore::ImageLoader::updateFromElement):

Canonical link: <a href="https://commits.webkit.org/269243@main">https://commits.webkit.org/269243@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aeda18b8cff123232ccceb9e66b7111bf566b45b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21901 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22125 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22967 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23782 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20280 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22147 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26369 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22435 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21378 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22132 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21761 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18990 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24634 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18898 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19843 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26115 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19924 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20064 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23974 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20511 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17477 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19856 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5244 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24062 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20456 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/ImageLoader.h:
```diff
@@ -117,6 +118,8 @@ class ImageLoader : public CachedImageClient {
 
     VisibleInViewportState imageVisibleInViewport(const Document&) const override;
 
+    bool canReuseFromListOfAvailableImages(const CachedResourceRequest&, const String& crossOriginAttribute);
```
Can this be marked as `static`, if so, please do it. If it cannot, please consider marking it as `const` at least.

szewai commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -164,6 +165,22 @@ void ImageLoader::clearImageWithoutConsideringPendingLoadEvent()
+    if (resource->options().credentials == FetchOptions::Credentials::SameOrigin && (crossOriginAttribute.isNull() || equalLettersIgnoringASCIICase(crossOriginAttribute, "use-credentials"_s)))
+        return false;
+
+    return true;
```
Should it be false when `resource->options().credentials == FetchOptions::Credentials::Include` and `crossOriginAttribute` is anonymous?

pgorszkowski-igalia replied:
You are right, based on what is in https://html.spec.whatwg.org/multipage/urls-and-fetching.html#cors-settings-attribute it should be added too. In case `resource->options().credentials == FetchOptions::Credentials::Include`, `crossOriginAttribute` can't be null or different from `use-credentials`. I will add it.

pgorszkowski-igalia replied:
It seems that by default `resource->options().credentials` is set to `FetchOptions::Credentials::Include`:
https://github.com/WebKit/WebKit/blob/main/Source/WebCore/loader/ImageLoader.cpp#L187
https://github.com/WebKit/WebKit/blob/main/Source/WebCore/loader/cache/CachedResourceLoader.cpp#L1793

If the `crossorigin` attribute is not specified, `resource->options().credentials` should be set to `FetchOptions::Credentials::Omit` in: 
https://github.com/WebKit/WebKit/blob/main/Source/WebCore/loader/CrossOriginAccessControl.cpp#L162

In other case, we have situation (if `crossorigin` is not specified at all) that:
`resource->options().credentials == FetchOptions::Credentials::Include` and `crossOriginAttribute` is null

WDYT?

pgorszkowski-igalia replied:
If I set `FetchOptions::Credentials::Omit` for `resource->options().credentials`  in case of `crossOriginAttribute` is null than few tests fail. The reason for fails is that tests are expected to include credentials in the request of the resource.

szewai replied:
I think the problem is, when we decide if there is a match in `crossorigin` attribute of request and cached resource,
we only check if credentials mode (https://fetch.spec.whatwg.org/#concept-request-credentials-mode) is expected, but not mode (https://fetch.spec.whatwg.org/#concept-request-mode, `FetchOptions::mode`). According to spec of creating a core request (https://html.spec.whatwg.org/multipage/urls-and-fetching.html#create-a-potential-cors-request), `crossorigin` attribute could affect both mode and credentials mode. 

For example, in the case where img element's `crossorigin` attribute is not specified, potential request credentials mode should be 'include' and mode should be 'no-cors'. My understanding is if the cached resource's credentials mode is 'include' and mode is 'cors', it's a mismatch.

pgorszkowski-igalia replied:
Ok, now I see your point. Maybe I should also add tests similar to `LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/image-loading-lazy-use-list-of-available-images.html` but add options with different values for `crossorigin`.

szewai commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -110,6 +111,28 @@ static inline bool pageIsBeingDismissed(Document& document)
 }
 
+// https://html.spec.whatwg.org/multipage/images.html#updating-the-image-data:list-of-available-images
+static bool canReuseFromListOfAvailableImages(const CachedResourceRequest& request, const String& crossOriginAttribute, Document& document)
```
Maybe we can just compare `request.options()` and `resource.options()` since `request` is created based on `crossOriginAttribute`, i.e. there is no need to pass in `crossOriginAttribute`.

pgorszkowski-igalia replied:
It seems to be a perfect solution:)

pgorszkowski-igalia committed:
Implement list of available images matching logic
 https://bugs.webkit.org/show_bug.cgi?id=243790

Reviewed by Sihui Liu and Chris Dumez.

Implement list of available images matching logic:
https://html.spec.whatwg.org/multipage/images.html#updating-the-image-data:list-of-available-images

This is restricted to lazy loading for now.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/image-loading-lazy-available-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/image-loading-lazy-use-list-of-available-images-expected.txt:
* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::canReuseFromListOfAvailableImages):
(WebCore::ImageLoader::updateFromElement):

Canonical link: https://commits.webkit.org/269243@main
